language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -346,96 +355,93 @@ impl AbstractState {
-                    references.insert(idx.clone(), nonce);
+    pub fn join_(mut self, mut other: Self) -> Option<Self> {
+        checked_precondition!(self.is_canonical() && other.is_canonical());
+        checked_precondition!(self.next_id == other.next_id);
+        checked_precondition!(self.frame_root == other.frame_root);
+        // A join failure (returning None) occurs in each of the following situations:
+        // - a resource is available along one path but not the other
+        // - a value is borrowed along one path but unavailable along the other
+        let local_idxs = {
+            let mut idxs = self.locals.keys().cloned().collect::<BTreeSet<_>>();
+            idxs.extend(other.locals.keys().cloned());
+            idxs
+        };
+        let mut locals = BTreeMap::new();
+        for idx in &local_idxs {
+            let self_value = self.locals.get(idx);
+            let other_value = other.locals.get(idx);
+            checked_verify!(self_value.is_some() || other_value.is_some());
+            match (self_value, other_value) {
+                // join error, a resource is available along one path but not the other
+                (None, Some(v)) |
+                (Some(v), None) if v.value.is_possibly_resource() => return None,
+
+                // A reference is assigned on one side, but not the other. Release
+                (None, Some(v)) => {
+                    if let AbstractValue::Reference(_) = &v.value {
+                        other.destroy_local(*idx);
+                    }
+                }
+                (Some(v), None) => {
+                    if let AbstractValue::Reference(_) = &v.value {
+                        self.destroy_local(*idx);
+                    }
                 }
-                AbstractValue::Value(kind) => {
-                    values.insert(idx.clone(), kind);
+
+                // The local has a value on each side, add it to the state
+                (Some(v1), Some(v2)) => {
+                    checked_verify!(v1 == v2);
+                    checked_verify!(!locals.contains_key(idx));
+                    locals.insert(*idx, v1.clone());
                 }
+
+                (None, None) => unreachable!("The local indexes are a union of the two keys. The local must be present in at least one map"),
             }
         }
+
+        // Join the borrow graph after releasing
+        let borrow_graph = {
+            self.borrow_graph.join(&other.borrow_graph);
+            self.borrow_graph
+        };
+        let next_id = self.next_id;
+        let frame_root = self.frame_root;
+        let joined = Self {
+            locals,
+            borrow_graph,
+            next_id,
+            frame_root,
+        };
+
+        // If there is a reference that is borrowing an empty local, join error
+        let borrowed_value_unavailable = local_idxs
+            .iter()
+            .any(|idx| joined.is_local_borrowed(*idx) && !joined.locals.contains_key(idx));
```

Comment:
Is there a similar problem where a reference is borrowing a global that is `move_from`'d on one control-flow path, but not the other?

Reply:
I believe the answer is no. 

We only care about `move_from` while there is an outstanding borrow. If it is moved in one branch and not the other, that does not interfere with any borrows created in that state (as each branch is independent)

Reply:
Hmm, I think I'm still confused. Would you mind pointing me to the test that triggers a join failure in this code? I'll try to construct the equivalent example with global references and play around with it locally to help me understand.

Reply:
The test you commented on fails because of this check:
language/functional_tests/tests/testsuite/commands/join_failure.mvir

But this check is going away shortly